include/eve/detail/function/simd/common/canonical_shuffle_adapter.hpp

```diff
@@ -135,96 +135,125 @@ upscale_pattern(pattern_t<I...> p)
 {
   // On an emulation bitcasting types won't be helpful for anything.
-  if constexpr( !supports_simd ) { return simplified_pattern {x, g, p}; }
+  if constexpr( !supports_simd ) { return simplified_pattern {kumi::tuple {x, xs...}, g, p}; }
   else if constexpr( G >= 2 && sizeof(T) < 8 )
   {
-    if constexpr ( sizeof(T) == 4 && current_api == neon)
+    if constexpr( sizeof(T) == 4 && current_api == neon )
     {
-      using up_t = std::uint64_t;
-      const auto up = bit_cast(x, as<wide<up_t, typename N::split_type>>());
-      return simplify_plain_shuffle1(up, eve::lane<G / 2>, p);
+      using up_t = eve::wide<std::uint64_t, typename N::split_type>;
+      return simplify_plain_shuffle1(
+          p, eve::lane<G / 2>, bit_cast(x, as<up_t> {}), bit_cast(xs, as<up_t> {})...);
     }
     else
     {
-      using up_t    = upgrade_t<T>;
-      const auto up = bit_cast(x, as<wide<up_t, typename N::split_type>>());
-      return simplify_plain_shuffle1(up, eve::lane<G / 2>, p);
+      using up_t = eve::wide<upgrade_t<T>, typename N::split_type>;
+      return simplify_plain_shuffle1(
+          p, eve::lane<G / 2>, bit_cast(x, as<up_t> {}), bit_cast(xs, as<up_t> {})...);
     }
   }
   else if constexpr( std::signed_integral<T> )
   {
-    using u_t = std::make_unsigned_t<T>;
-    auto cast = bit_cast(x, as<wide<u_t, N>>());
-    return simplify_plain_shuffle1(cast, g, p);
+    using u_t = eve::wide<std::make_unsigned_t<T>, N>;
+    return simplify_plain_shuffle1(p, g, bit_cast(x, as<u_t> {}), bit_cast(xs, as<u_t> {})...);
   }
   else if constexpr( current_api == avx && std::same_as<T, std::uint32_t> && N::value == 8 )
   {
-    auto cast = bit_cast(x, as<wide<float, N>> {});
-    return simplify_plain_shuffle1(cast, g, p);
+    using f_t = eve::wide<float, N>;
+    return simplify_plain_shuffle1(p, g, bit_cast(x, as<f_t> {}), bit_cast(xs, as<f_t> {})...);
   }
   else if constexpr( current_api == avx && std::same_as<T, std::uint64_t> && N::value == 4 )
   {
-    auto cast = bit_cast(x, as<wide<double, N>> {});
-    return simplify_plain_shuffle1(cast, g, p);
+    using f_t = eve::wide<double, N>;
+    return simplify_plain_shuffle1(p, g, bit_cast(x, as<f_t> {}), bit_cast(xs, as<f_t> {})...);
   }
   else if constexpr( fundamental_cardinal_v<T> > N::value )
   {
-    auto cast         = bit_cast(x, as<wide<T, fundamental_cardinal_t<T>>> {});
     auto next_pattern = []<std::size_t... i>(std::index_sequence<i...>) {
       return pattern_t<I..., ((void)i, we_)...> {};
     }(std::make_index_sequence<fundamental_cardinal_v<T> - sizeof...(I)> {});
-    return simplify_plain_shuffle1(cast, g, next_pattern);
+    using T1 = wide<T, fundamental_cardinal_t<T>>;
+    return simplify_plain_shuffle1(
+        next_pattern, g, bit_cast(x, as<T1> {}), bit_cast(xs, as<T1> {})...);
   }
-  else { return simplified_pattern {x, g, p}; }
+  else { return simplified_pattern {kumi::tuple {x, xs...}, g, p}; }
```

Comment:
I should add the inversion somewhere. So that `0, 5, 2, 3' and '4, 1, 6, 7' have to be processed once.